### ISSUE TITLE
DualLink fails on Variblenode

### ISSUE THREAD

**Issue by noskill** [opencog/atomspace#2296]
Modified example from  https://wiki.opencog.org/w/DualLink

```scheme
(BindLink
  (ListLink
     (Concept "I") (Glob "$star") (Concept "you"))
  (ListLink
     (Concept "I") (Glob "$star") (Concept "you") (Concept "too")))
```

```scheme
(cog-execute! (DualLink                        
   (ListLink (Concept "I") (VariableNode "$star") (Concept "you"))))               
```
$3 = (SetLink
)

While BindLink works as expected
```scheme
(cog-execute! (BindLink
  (ListLink
     (Concept "I") (Glob "$star") (Concept "you"))
  (ListLink
     (Concept "I") (Glob "$star") (Concept "you") (Concept "too"))))
```

$5 = (SetLink
   (ListLink
      (ConceptNode "I")
      **(VariableNode "$star")**
      (ConceptNode "you")
      (ConceptNode "too")
   )
)

**Comment by linas** [comment]
I'm strongly tempted to close this as not-a-bug, working as designed. What answer were you expecting to get? Why do you think this is a bug?

**Comment by linas** [comment]
This works for me:
```
(use-modules (opencog))
(use-modules (opencog exec))

 (BindLink
  (ListLink
     (Concept "I") (Glob "$star") (Concept "you"))
  (ListLink
     (Concept "I") (Glob "$star") (Concept "you") (Concept "too")))

(cog-execute! (DualLink                        
   (ListLink (Concept "I") (GlobNode "$star") (Concept "you")))) 
```
returns exactly what is expected:
```
 
$3 = (SetLink
   (ListLink
      (ConceptNode "I")
      (VariableNode "$star")
      (ConceptNode "you")
   )
)
```

**Comment by linas** [comment]
I remvoed the last comment, it was an incorrect cut and paste. Let me try again.

**Comment by linas** [comment]
This works for me, giving the expected answer:
```
(use-modules (opencog))
(use-modules (opencog exec))
 (ListLink (Concept "I") (VariableNode "$star") (Concept "you"))

(cog-execute! (DualLink                        
   (ListLink (Concept "I") (GlobNode "$star") (Concept "you"))))
```
returns:
```
$2 = (SetLink
   (ListLink
      (ConceptNode "I")
      (VariableNode "$star")
      (ConceptNode "you")
   )
)
```
which is expected.

**Comment by linas** [comment]
I edited the last comment for clarity; please view github and not email to see what I'm trying to say. Sorry for confusion.

**Comment by noskill** [comment]
So why i can't use ListLink with some variable node to match conditional part of the bindlink with another variable?
```scheme
(ListLink (Concept "I") (Variable "$planet") (Concept "you"))
```

$5 = (ListLink
   (ConceptNode "I")
   (VariableNode "$planet")
   (ConceptNode "you")
)

```scheme
scheme@(guile-user)> (cog-execute! (BindLink   (ListLink 
     (Concept "I") (Variable "$star") (Concept "you"))
  (ListLink
     (Concept "I") (Variable "$star") (Concept "you") (Concept "too"))))
```
$6 = (SetLink
   (ListLink
      (ConceptNode "I")
      (VariableNode "$planet")
      (ConceptNode "you")
      (ConceptNode "too")
   )
)
```scheme
(cog-execute! (DualLink                        
   (ListLink (Concept "I") (VariableNode "$planet") (Concept "you"))))         
```      
$9 = (SetLink
)

**Comment by noskill** [comment]
I expect duallink be dual to bindlink, i.g. to return condition of the bindlink given the data, if the data matches in bindlink execution.

**Comment by linas** [comment]
OK, Yes, that kind-of makes sense.

Historically, when DualLink was created, it did not know about Variables; it used only Globs.

Since then, it was noted that a Variable is "just like a Glob of length one".  Which does seem reasonable, I guess. It's hard to see why it would not be.   However, there are many places in the code which do NOT treat globs and variables as (almost) interchangeable things.  To fix it, all of those places would need to be found and updated. Large changes like this make me slightly nervous.

So, today, one can use Globs to find Variables, but one mostly cannot use Variables to find other Variables. And for most users, this seems to be enough.  I guess we could change this.... 

...but I'm happy to keep things as they are.   I don't like change-for-the-sake-of-change; I tend to like keeping things as they are, it helps with stability.
